src/lib/index.ts

```diff
@@ -0,0 +1,277 @@
+      // By removing it from the last position and adding it back after, we ensure
+      // it keeps this position.
+      // fixes #3148
+      const lastWidget = localWidgets.pop();
+
+      localWidgets = localWidgets.concat(widgets);
+
+      if (lastWidget) {
+        // Second part of the fix for #3148
+        localWidgets = localWidgets.concat(lastWidget);
+      }
+
+      if (localInstantSearchInstance && Boolean(widgets.length)) {
+        helper!.setState(
+          localWidgets.reduce(enhanceConfiguration, {
+            ...helper!.state,
+          })
+        );
+
+        widgets.forEach(widget => {
+          if (localInstantSearchInstance && widget.init) {
+            widget.init({
+              helper: helper!,
+              parent: this,
+              instantSearchInstance: localInstantSearchInstance,
+              state: helper!.state,
+              templatesConfig: localInstantSearchInstance.templatesConfig,
+              createURL: localInstantSearchInstance._createAbsoluteURL,
+              onHistoryChange: localInstantSearchInstance._onHistoryChange,
+            });
+          }
+        });
+
+        localInstantSearchInstance.scheduleSearch();
+      }
+
+      return this;
+    },
+
+    removeWidgets(widgets) {
+      if (!Array.isArray(widgets)) {
+        throw new Error(
+          withUsage('The `removeWidgets` method expects an array of widgets.')
+        );
+      }
+
+      if (widgets.some(widget => typeof widget.dispose !== 'function')) {
+        throw new Error(
+          withUsage('The widget definition expects a `dispose` method.')
+        );
+      }
+
+      localWidgets = localWidgets.filter(
+        widget => widgets.indexOf(widget) === -1
+      );
+
+      if (localInstantSearchInstance && widgets.length) {
+        const nextState = widgets.reduce((state, widget) => {
+          // the `dispose` method exists at this point we already assert it
+          const next = widget.dispose!({ helper: helper!, state });
+
+          return next || state;
+        }, helper!.state);
+
+        helper!.setState(
+          localWidgets.reduce(enhanceConfiguration, {
+            ...nextState,
+          })
+        );
+
+        if (localWidgets.length) {
+          localInstantSearchInstance.scheduleSearch();
+        }
+      }
+
+      return this;
+    },
+
+    init({ instantSearchInstance, parent }) {
+      localInstantSearchInstance = instantSearchInstance;
+
+      // The `mainHelper` is already defined at this point. The instance is created
+      // inside InstantSerach at the `start` method, which occurs before the `init`
+      // step.
+      const mainHelper = instantSearchInstance.mainHelper!;
+
+      const initialSearchParameters =
+        // Uses the `searchParameters` for the top level index only, it allow
+        // us to have the exact same behaviour than before for the mono-index.
+        parent === null ? instantSearchInstance._searchParameters : {};
```

Comment:
The support of `searchParameters` is (still) required for default values. The `configure` widget does not behave exactly the same when a widget is added or removed. If the default value is removed by the user with a `configure`  widget (+ routing) the filters are re-applied on add/remove.